chromium_src/chrome/browser/notifications/notification_display_service_impl.cc

```diff
@@ -0,0 +1,130 @@
+  }
+}
+
+void NotificationDisplayServiceImpl::ShutdownTimerCallback() {
+  // Check if browser runs without UI and close it if so. If during
+  // shutdown a new notification event arrives, timer is stopped in
+  // ProcessNotificationOperation
+  auto state = base::android::ApplicationStatusListener::GetState();
+  bool no_ui =
+      (state == base::android::ApplicationState::APPLICATION_STATE_UNKNOWN ||
+       state == base::android::ApplicationState::
+                    APPLICATION_STATE_HAS_DESTROYED_ACTIVITIES);
+  if (no_ui) {
+    chrome::TerminateAndroid();
+  }
+}
+
+void NotificationDisplayServiceImpl::ProcessNotificationOperation(
+    NotificationCommon::Operation operation,
+    NotificationHandler::Type notification_type,
+    const GURL& origin,
+    const std::string& notification_id,
+    const base::Optional<int>& action_index,
+    const base::Optional<base::string16>& reply,
+    const base::Optional<bool>& by_user) {
+  NotificationHandler* handler = GetNotificationHandler(notification_type);
+  DCHECK(handler);
+  if (!handler) {
+    LOG(ERROR) << "Unable to find a handler for "
+               << static_cast<int>(notification_type);
+    return;
+  }
+
+  no_ui_shutdown_timer_.Stop();
+
+  base::OnceClosure completed_closure =
+      base::BindOnce(&NotificationDisplayServiceImpl::BraveOperationCompleted,
+                     base::Unretained(this), notification_id);
+  base::OnceClosure notification_handler;
+
+  switch (operation) {
+    case NotificationCommon::OPERATION_CLICK:
+      notification_handler = base::BindOnce(
+          &NotificationHandler::OnClick, base::Unretained(handler), profile_,
+          origin, notification_id, action_index, reply,
+          std::move(completed_closure));
+      break;
+    case NotificationCommon::OPERATION_CLOSE:
+      DCHECK(by_user.has_value());
+      // Save last dismissed notification ID
+      last_dismissed_notification_id_ = notification_id;
+      notification_handler = base::BindOnce(
+          &NotificationHandler::OnClose, base::Unretained(handler), profile_,
+          origin, notification_id, by_user.value(),
+          std::move(completed_closure));
+      break;
+    case NotificationCommon::OPERATION_DISABLE_PERMISSION:
+      notification_handler =
+          base::BindOnce(&NotificationHandler::DisableNotifications,
+                         base::Unretained(handler), profile_, origin);
+      break;
+    case NotificationCommon::OPERATION_SETTINGS:
+      notification_handler =
+          base::BindOnce(&NotificationHandler::OpenSettings,
+                         base::Unretained(handler), profile_, origin);
+      break;
+  }
+
+  if (brave_ads_initialized_ && !notification_handler.is_null()) {
+    std::move(notification_handler).Run();
+  } else {
+    brave_ads_actions_.push(std::move(notification_handler));
+  }
+}
+
+void NotificationDisplayServiceImpl::OnBraveAdsServiceReady(bool success) {
```

Comment:
I don't think the notification display service should have to know whether brave ads is running or not